client/my-sites/plans-features-main/hooks/use-action-callback.ts

```diff
@@ -0,0 +1,188 @@
+
+function useUpgradeHandler(
+	siteSlug?: string | null,
+	withDiscount?: string,
+	cartHandler?: ( cartItems?: MinimalRequestCartProduct[] | null ) => void
+) {
+	// TODO:
+	// - clickedPlanSlug can likely be removed
+	const processCartItems = useCallback(
+		( cartItems?: MinimalRequestCartProduct[] | null ) => {
+			const cartItemForPlan = getPlanCartItem( cartItems );
+
+			const cartItemForStorageAddOn = cartItems?.find(
+				( items ) => items.product_slug === PRODUCT_1GB_SPACE
+			);
+
+			if ( cartItemForStorageAddOn?.extra ) {
+				recordTracksEvent( 'calypso_signup_storage_add_on_upgrade_click', {
+					add_on_slug: cartItemForStorageAddOn.extra.feature_slug,
+				} );
+			}
+
+			if ( cartHandler ) {
+				cartHandler( cartItems );
+				return;
+			}
+
+			const planPath = cartItemForPlan?.product_slug
+				? getPlanPath( cartItemForPlan.product_slug )
+				: '';
+
+			const checkoutUrl = cartItemForStorageAddOn
+				? `/checkout/${ siteSlug }/${ planPath },${ cartItemForStorageAddOn.product_slug }:-q-${ cartItemForStorageAddOn.quantity }`
+				: `/checkout/${ siteSlug }/${ planPath }`;
+
+			const checkoutUrlWithArgs = addQueryArgs(
+				{ ...( withDiscount && { coupon: withDiscount } ) },
+				checkoutUrl
+			);
+
+			page( checkoutUrlWithArgs );
+			return;
+		},
+		[ siteSlug, withDiscount, cartHandler ]
+	);
+
+	return useCallback(
+		( {
+			cartItemForPlan,
+			isFreeTrialCta,
+			selectedStorageAddOn,
+		}: {
+			cartItemForPlan?: MinimalRequestCartProduct | null;
+			isFreeTrialCta: boolean | undefined;
+			selectedStorageAddOn?: AddOns.AddOnMeta | null;
+		} ) => {
+			/* 1. Process free trial plan for checkout. Exclude add-ons. */
+			if ( isFreeTrialCta && cartItemForPlan ) {
+				processCartItems?.( [ cartItemForPlan ] );
+				return;
+			}
+
+			/* 2. Process plan and add-ons for checkout. */
+			const storageAddOnCartItem = selectedStorageAddOn &&
+				! selectedStorageAddOn.purchased && {
+					product_slug: selectedStorageAddOn.productSlug,
+					quantity: selectedStorageAddOn.quantity,
+					volume: 1,
+					extra: { feature_slug: selectedStorageAddOn?.featureSlugs?.[ 0 ] },
+				};
+
+			if ( cartItemForPlan ) {
+				processCartItems?.( [
+					cartItemForPlan,
+					...( storageAddOnCartItem ? [ storageAddOnCartItem ] : [] ),
+				] );
+				return;
+			}
+
+			/* 3. Process free plan for checkout */
+			processCartItems?.( null );
+			return;
+		},
+		[ processCartItems ]
+	);
+}
+
+function useGenerateActionCallback(
+	// TODO: Reevaluate param types and whether or not they should be optional
+	currentPlan: Plans.SitePlan | undefined,
+	eligibleForFreeHostingTrial: boolean,
+	cartHandler?: ( cartItems?: MinimalRequestCartProduct[] | null ) => void,
+	flowName?: string | null,
+	intent?: PlansIntent | null,
+	planActionCallback?: ( planSlug: PlanSlug ) => boolean,
+	sitePlanSlug?: PlanSlug | null,
+	siteSlug?: string | null,
+	withDiscount?: string
+): UseActionCallback {
```

Comment:
Same. Can we convert this to an object prop (`{ currentPlan, cartHandler, etc. }`)?

Reply:
Yeah we totally can -- I'm on it 👍

Reply:
https://github.com/Automattic/wp-calypso/pull/88694/commits/9989ba040c427175bb136b39721e1acd998f0db6